Prevent equivocation dos

gterzian opened the pull request:
FIX #873

gterzian commented:
@rphmeier Could you take a look and tell me if this is going in the right direction in terms of biz logic?


Still a few TODOS:
1. Write a test
2. Figure out if some of the intermittently long-running tests in `substrate-finality-grandpa` have something to do with those changes(it appears the logic I added isn't actually hit in those tests)

rphmeier commented on core/finality-grandpa/src/communication/mod.rs:
```diff
@@ -273,6 +296,11 @@ impl<B: BlockT, N: Network<B>> NetworkBridge<B, N> {
 
 		let topic = round_topic::<B>(round.0, set_id.0);
+
+        let mut votes_tally: HashMap<AuthorityId, VoteTally> = HashMap::new();
```
spaces

pepyakin replied:
(not only here and not only in this file)

rphmeier commented on core/finality-grandpa/src/communication/mod.rs:
```diff
@@ -273,6 +296,11 @@ impl<B: BlockT, N: Network<B>> NetworkBridge<B, N> {
+
+        let mut votes_tally: HashMap<AuthorityId, VoteTally> = HashMap::new();
+		let num_signers = voters.voters().iter().count();
+		let max_timed_out = num_signers.checked_div(3).unwrap_or(1);
```
`finality_grandpa` exposes a `fn threshold` for doing this calculation. The other thing is that voters are weighted, so we should be doing it based on total timed-out weight as opposed to a total number of voters timed-out.

rphmeier commented:
On the direction:
  As long as we can ignore messages that are irrelevant, the time-out stuff is far less important and I wouldn't worry about implementing it until the ignoring stuff is done.

  I think a lot of this stuff needs to be done in the `communication/gossip` module, where we track the things that each peer has sent us and what we've sent a peer.  The `validate_message` function should drop any redundant equivocations (3rd vote or after) and punish any peer who sends us a redundant equivocation _when they should have known not to_. "oblivious" misbehavior is fine.

gterzian commented:
@rphmeier Ok, thanks for the additional info. I've had a chance to better look at the various parts of finality-grandpa and how those integrate with consensus gossip in network. 

Have you considered that it might be better to implement this logic at the level of the stream of incoming  and outgoing messages, with perhaps bits of highly targeted data shared between grandpa and network? 

For example re counting votes from incoming messages, we could add the proposed capabilities to `GossipValidator`, and then use it as part of [`on_incoming`](https://github.com/paritytech/substrate/blob/c26b54b5b4e2f70c148e6225cd1703f84b33eb4f/core/network/src/consensus_gossip.rs#L396), or we could just do that when the incoming message is handled as part of the stream made available by [`messages_for`](fhttps://github.com/paritytech/substrate/blob/c26b54b5b4e2f70c148e6225cd1703f84b33eb4f/core/finality-grandpa/src/communication/mod.rs#L278)? 

In some ways it's even those entire [`Validator`](https://github.com/paritytech/substrate/blob/c26b54b5b4e2f70c148e6225cd1703f84b33eb4f/core/network/src/consensus_gossip.rs#L201) and [`Network`](https://github.com/paritytech/substrate/blob/c26b54b5b4e2f70c148e6225cd1703f84b33eb4f/core/finality-grandpa/src/communication/mod.rs#L278) traits that could be removed, and replaced with a combination of what's already there in those streams, and perhaps some more targeted shared data with `ConsensusGossip`. 

If we're worried about the storage costs of messages, I think the most obvious thing to remove is this [`Vec<MessageEntry<B>>`](https://github.com/paritytech/substrate/blob/c26b54b5b4e2f70c148e6225cd1703f84b33eb4f/core/network/src/consensus_gossip.rs#L228) on `ConsensusGossip`.  It appears to me that the content of this vec are mostly used to pass stuff back to `GossipValidator` via method calls... As to the actual messages that have to be gossiped out over the network, could those not just come out of [`OutoingMessages`](fhttps://github.com/paritytech/substrate/blob/c26b54b5b4e2f70c148e6225cd1703f84b33eb4f/core/finality-grandpa/src/communication/mod.rs#L278), or otherwise be made available by grandpa on a "validated" basis? 

For example, why does [`gossip_message`](https://github.com/paritytech/substrate/blob/c26b54b5b4e2f70c148e6225cd1703f84b33eb4f/core/finality-grandpa/src/communication/mod.rs#L278) ends ups calling [`multicast`](https://github.com/paritytech/substrate/blob/c26b54b5b4e2f70c148e6225cd1703f84b33eb4f/core/network/src/consensus_gossip.rs#L490), which then ends-up calling back into grandpa via [`message_allowed`](https://github.com/paritytech/substrate/blob/5f0c38086eb3fae4374d87db2a6abc2fbd270f48/core/finality-grandpa/src/communication/gossip.rs#L700). Could we not transform the stream of outgoing messages so that it only contains messages that are "allowed", and have network gossip them when they are received over a stream without the need to checking with the `GossipValidator` whether they are allowed?

Another example, looking at [`on_incoming`](https://github.com/paritytech/substrate/blob/c26b54b5b4e2f70c148e6225cd1703f84b33eb4f/core/network/src/consensus_gossip.rs#L396). 
1. First, we [`validate`](https://github.com/paritytech/substrate/blob/c26b54b5b4e2f70c148e6225cd1703f84b33eb4f/core/network/src/consensus_gossip.rs#L416),  the message, using the gossip validator, and then we handle the result of this validation. 
2. Then, if the validation passes, we enqueue the message onto the `live_message_sinks`, and then in some cases also store it in `messages`. 
3. Later, this message will be handled as part of the [stream of incoming messages for a round](https://github.com/paritytech/substrate/blob/c26b54b5b4e2f70c148e6225cd1703f84b33eb4f/core/finality-grandpa/src/communication/mod.rs#L276). 

There are a couple of problems with this: 

1. Double storage, inside [`messages`](https://github.com/paritytech/substrate/blob/c26b54b5b4e2f70c148e6225cd1703f84b33eb4f/core/network/src/consensus_gossip.rs#L228) of `ConsensusGossip`, and inside the stream of round communication. 
2. Double decoding, once as part of [`validate`](https://github.com/paritytech/substrate/blob/5f0c38086eb3fae4374d87db2a6abc2fbd270f48/core/finality-grandpa/src/communication/gossip.rs#L626), and then later as part of [handling incoming messages on the stream](https://github.com/paritytech/substrate/blob/c26b54b5b4e2f70c148e6225cd1703f84b33eb4f/core/finality-grandpa/src/communication/mod.rs#L278).

Besides this, the biggest problem is that we end-up sharing this entire `GossipValidator` with `ConsensusGossip`, whereas I think we could just `validate` incoming messages as part of the stream from `messages_for`. In other words, why share the `GossipValidator` so that we can call this `validate` method, when we could do the equivalent later inside grandpa?

Even more complicated logic, such as that found in [`propagate`](https://github.com/paritytech/substrate/blob/c26b54b5b4e2f70c148e6225cd1703f84b33eb4f/core/network/src/consensus_gossip.rs#L396) could be implemented without sharing the entire `GossipValidator`. 

For example I think the lock is held for way too long as part of the `message_allowed` flow, and the fact that it's a read lock doesn't really help since it will still block other write accesses, of which there are plenty in grandpa. 

 Instead of a call to a method of `GossipValidator` like `message_allowed` there, a piece of targeted shared data could do the trick. The entire logic run [here](https://github.com/paritytech/substrate/blob/5f0c38086eb3fae4374d87db2a6abc2fbd270f48/core/finality-grandpa/src/communication/gossip.rs#L719) could be encoded in a shared data structure and updated by grandpa on a continuous basis. From a quick look, it would appear to me that something like a `HashMap<PeerId, HashSet<topic>>`, together with a `boolean` for `do_rebroadcast` and some info about `last_commit`, could do the trick. Such a data-structure would be updated by `GossipValidator` whenever "something changes", and `ConsensusGossip` could read it whenever it needs to know whether a "message is allowed". The lock would probably only rarely be contented, since it would only be held for the actual updating and reading of the data, as opposed to for the duration of running some "message allowed" logic. 

It would be even better if `ConsensusGossip` didn't have any concept of "message allowed", and that it would broadcast stuff coming from an already validated stream of outgoing messages from grandpa. 

A last example: things like [`rebroadcast`](https://github.com/paritytech/substrate/blob/c26b54b5b4e2f70c148e6225cd1703f84b33eb4f/core/network/src/consensus_gossip.rs#L317) and [`broadcast_topic`](https://github.com/paritytech/substrate/blob/c26b54b5b4e2f70c148e6225cd1703f84b33eb4f/core/network/src/consensus_gossip.rs#L324), currently operate on messages stored at the network layer, and then ask grandpa to validate each message before sending them out. Why not just have a shared map of messages that are updated by grandpa on an ongoing basis, and available for network to broadcast when necessary? It would allow logic to run in parallel, with quick locking and writing/reading of shared-data when necessary, as opposed to just lock the entire inner state of `GossipValidator` while the logic is run. 

It appears to me that currently we're spreading logic around, and possible doing a lot of extra work, and not getting a lot of parallelism on those operations. I'd argue for doing all that validation stuff as part of stream handling in grandpa where possible, just provide a stream of "validated" messages for network to sent out, and if absolutely necessary share some data in a very targeted way.

rphmeier commented:
I haven't had time to read the whole post, but I think that it isn't possible to handle all of the gossip logic purely at the incoming stream / outgoing sink abstraction. That's a helpful abstraction to make things easier from the perspective of a specific round of GRANDPA, but at the implementation layer (gossip) it's not that simple.

We aren't really worried about the storage costs of messages _provided that there is a sensible bound on them_. Sure, the way that we store them right now isn't as optimal as it could be, but even if we could reduce that by 10x, what matters is how many messages we actually need to store at any moment. That's the problem this PR is meant to address (algorithmic improvement, not micro-optimization).

>  I think we could validate incoming messages as part of the stream from messages_for

Message validation is heavily dependent on which peer the message comes from, and what data we've sent to them.

> Could we not transform the stream of outgoing messages so that it only contains messages that are "allowed", and have network gossip them when they are received over a stream without the need to checking with the GossipValidator whether they are allowed

I think this misses the point. Obviously messages are not blanket allowed/disallowed, but it is dependent on the protocol state of the peer which we are sending to. `fn message_allowed` is meant to provide logic for evaluating whether we should send a message to a _specific_ peer based on the information we have about this peer.

gterzian commented:
Ok, thanks for the additional info. 

So if it's not possible to just do everything with a stream of incoming/outgoing messages, as exists for voting rounds, perhaps because things like "validation" and "expiration" of messages doesn't tie in well with the lifetime of a voting round, I think we should rename the `GossipValidator` into something like `GossipeMessageCache`, and move all the message storage into it, and merge it with the validation logic. 

Things like the vec of "messages" found on network, combined with a stream of basically the same messages being handled by grandpa, isn't so much a concern regarding storage costs, the problem is rather that logic, and data, is spread across various places. It could make sense to consolidate it. 

If I look at `on_incoming`, `collect_garbage` or `propagate` in gossip, it appears to me that the messages stored on gossip are then passed to the "validator" for some processing, and then the result tells gossip what to do with the message. 

If we instead store all messages inside the validator, which I'd rather call a "cache", then gossip could just do things like: 

1. collect garbage: `let still_good = cache.revalidate(message_hash)`, where hash comes from `known_messages`. 
2. propagate a message: `let message = cache.get_message(message_hash)`, if `None` is returned, we know the message is "not allowed" to be sent out. 
3. on incoming `let hash = cache.validate(message)`, with gossip only storing `Some(hash)` if one is returned by the cache. 

And coming back to the sub-optimal locking in `message_allowed`, instead of more granular locking, the solution could be merging storage and validation into one place, basically a cache. That would be folding the logic into data, moving it away from procedural code that is currently found across gossip and grandpa.

rphmeier commented:
Sounds like a reasonable set of refactorings. No, I don't view refactoring as optimization, and certainly the duplication of decoding is something we've grumbled about in the past. Folding the message storage into the `GossipValidator` is fine (and we can bikeshed on the name), although we probably want to store two copies: one encoded (for fast sending via simple `Clone`) and one decoded (for fast analysis relative to a peer's view). 

The convenient aspect of having the `ConsensusGossip` object manage all storage and track which peers have seen what is that we don't then have to duplicate this logic in all implemented gossip validators -- this is something we would have to avoid when doing that refactor.

gterzian commented:
@rphmeier Ok so I've reworked this to a more simple logic happening inside `gossip`, currently it would start rejecting messages from a peer for a given round, if that peer has sent us already 2 messages of a given type. There is no concept of time-out, or of ensuring we still accept enough messages to continue. 

How is it looking so far? 

Re the rest of our discussion, I'll look at addressing some of those points in follow-ups...

niklasad1 commented on core/finality-grandpa/src/communication/gossip.rs:
```diff
@@ -94,6 +95,25 @@ enum Consider {
+    RejectEquivocation,
+}
+
+#[derive(Debug)]
```
why not derive `Default`?

gavofyork commented:
@gterzian build failing? @niklasad1 @rphmeier any further comments or good to merge?

gterzian commented:
From what I can tell, the build failure is an [unrelated CI system failure](https://gitlab.parity.io/parity/substrate/-/jobs/151328)(have you noticed @gabreal ?).

I think this is still a work in progress, and I would like some intermediary feedback on the approach taken, @rphmeier

rphmeier commented on core/finality-grandpa/src/communication/gossip.rs:
```diff
@@ -94,6 +95,15 @@ enum Consider {
 	/// Message is from the future. Reject.
 	RejectFuture,
+    /// Message is an equivocation. Reject.
+    RejectEquivocation,
```
spaces

rphmeier replied:
I think this should also be `RejectRedundant`, because we do accept _some_ equivocations.

rphmeier commented on core/finality-grandpa/src/communication/gossip.rs:
```diff
@@ -94,6 +95,15 @@ enum Consider {
+}
+
+#[derive(Debug, Default)]
+pub(crate) struct VoteTally {
```
docs on public or pub(crate) items

rphmeier commented on core/finality-grandpa/src/communication/gossip.rs:
```diff
@@ -300,6 +310,8 @@ pub(super) enum Misbehavior {
 	// always misbehavior.
 	FutureMessage,
+    // Equivocation received
+    Equivocation,
```
(more spaces, here and elsewhere)

rphmeier replied:
also, sending an equivocation is _not_ misbehavior (sometimes, equivocations are required to complete a round). sending too many equivocations is misbehavior

gterzian replied:
Should I perhaps instead do something using the `_equivocation` "callbacks" on `Environment`, as opposed to determining that something is an equivocation just by counting incoming vote messages? 

It appears to me that the only way to know for sure if a message is indeed an equivocation is by using that handle to get the feedback from the actual grandpa worker.

Also, how do we determine if something is "too many" equivocations?

rphmeier replied:
Since a single equivocation is counted as all possible equivocations, we want to ignore all except the first because they are redundant.

We know something is an equivocation even without the callbacks because the definition is 2 messages by the same voter of the same type in the same round.

rphmeier commented on core/finality-grandpa/src/communication/gossip.rs:
```diff
@@ -476,7 +493,29 @@ impl<Block: BlockT> Inner<Block> {
 
-	fn consider_vote(&self, round: Round, set_id: SetId) -> Consider {
+	fn consider_vote(&mut self, round: Round, set_id: SetId, msg: &SignedMessage<Block>, id: &PeerId) -> Consider {
+		let mut tally = self.votes_tally.entry(id.clone()).or_insert(Default::default());
```
logic here is wrong -- this is tallying equivocations per _peer_ but note that peers in the gossip network are not the same as _voters_ who issue votes. Even if there are 10 voters, we could have 1000 peers.

rphmeier replied:
The other broken piece of logic here is that the vote tallies aren't scoped by round.

gterzian replied:
> this is tallying equivocations per peer but note that peers in the gossip network are not the same as voters who issue votes

It appeared to me that the code in `communication` was not aware of voters, just of peers, but I'll take another look and see if I can use voters here instead of peers. 

> the vote tallies aren't scoped by round

The tallies are reset there when a new round is started: https://github.com/paritytech/substrate/pull/2479/files#diff-1c4fd5ec8a616ff527309711747b39d3R468

Is that inacurate?

gabreal commented:
> From what I can tell, the build failure is an [unrelated CI system failure](https://gitlab.parity.io/parity/substrate/-/jobs/151328)(have you noticed @gabreal ?).

I re-triggered the run of the ci job and afaics the ci is working again as it should. There has been an outage due to changes in the docker hub configuration.

rphmeier commented on core/finality-grandpa/src/communication/gossip.rs:
```diff
@@ -449,6 +467,8 @@ impl<Block: BlockT> Inner<Block> {
 		self.local_view.round = round;
 		self.local_view.set_id = set_id;
+		// Reset the votes-tally for a new round.
+        self.votes_tally = Default::default();
```
1. spaces
2. we can run multiple rounds in parallel, so that doesn't work

rphmeier commented on core/finality-grandpa/src/communication/gossip.rs:
```diff
@@ -476,7 +496,29 @@ impl<Block: BlockT> Inner<Block> {
+				tally.handled_primary_proposals += 1;
+			},
+			Prevote(_prevote) => {
+				tally.handled_pre_votes += 1;
```
perhaps this and the others should be a saturating addition so we don't overflow somehow?

rphmeier commented on core/finality-grandpa/src/communication/mod.rs:
```diff
@@ -64,6 +64,8 @@ mod cost {
 	pub(super) const BAD_SIGNATURE: i32 = -100;
 	pub(super) const MALFORMED_COMMIT: i32 = -1000;
 	pub(super) const FUTURE_MESSAGE: i32 = -500;
+    pub(super) const EQUIVOCATION: i32 = 500;
```
why do you consider circulating equiocations as misbehavior? I don't.

rphmeier commented on core/finality-grandpa/src/communication/gossip.rs:
```diff
@@ -94,6 +98,16 @@ enum Consider {
 	RejectFuture,
+    /// Message is redundant. Reject.
+    RejectRedundant,
+	RejectEquivocation,
```
indentation is still wrong. I don't see the purpose of `RejectEquivocation`. can you explain?

gterzian replied:
> We do want to reject all redundant equivocations, but we only want to punish peers based on the combination of what they've sent to us

`RejectEquivocation` is meant for the "punish peers based on the combination of what they've sent to us" case, whereas `RejectRedundant` is meant to reject a redundant equivocation, without downgrading the peer. 

I guess the naming is off?

rphmeier replied:
Yeah, the naming is really off.

gterzian commented:
Ok, so the current logic operates at the level of "type of messages", rejecting any received in excess of two for a round, and reporting peers whom sent use such message and with whom we would have sent two such messages before. 

Question: once we get two messages of a given kind for a round, should we then reject not just messages of that kind, but any additional messages for that voter? And if we have sent two messages of a kind for a given round/voter to a peer, should we then report that peer for any messages sent in addition to it? 

So basically, do we scope it per message kind, or does two of a kind result in any additional message being rejected, regardless of kind?

Finally, this currently doesn't cover the Peer sends us Message A for a voter, we send Message A to peer for the same voter, should we then reject any other Message A for that voter, and should we report the peer for sending us such an additional message? Basically, this currently tallies what we send and what we received, but it doesn't cross-tally it. We might want to do it as well? The problem there might be that we can't know for sure if a peer received the message we have sent, right?

gterzian commented:
Ok so I went for the stricter "one equivocation makes any kind of additional message redundant", and for a slightly "refined" version to report peer that doesn't just count what we sent and they've sent us, bu rather counts the exchanges(So two messages, one sent by a peer, one by use to that peer, followed by a third from the peer, will report the peer).

I was also thinking that when it comes to what we send a peer, we can't know for sure they've actually received. So we might want to only report a peer based on what they sent us, ignoring what we sent them?

rphmeier commented on core/finality-grandpa/src/communication/gossip.rs:
```diff
@@ -174,13 +192,18 @@ impl<B: BlockT> KeepTopics<B> {
 	}
 
-	fn push(&mut self, round: Round, set_id: SetId) {
+    // Adds a round of topic, returns a list of pruned rounds
```
```suggestion
	// Adds a round of topic, returns a list of pruned rounds
```

rphmeier commented on core/finality-grandpa/src/communication/gossip.rs:
```diff
@@ -423,6 +453,9 @@ struct Inner<Block: BlockT> {
 	next_rebroadcast: Instant,
+	incoming_votes_tally: HashMap<Round, HashMap<AuthorityId, VoteTally>>,
+	incoming_msg_tally: HashMap<Round, HashMap<(AuthorityId, PeerId), VoteTally>>,
+	outgoing_votes_tally: HashMap<Round, HashMap<(AuthorityId, PeerId), VoteTally>>,
```
Keying these only by round is a little fragile. If we have 2 set changes shortly after another, we could have 2 rounds with the same number.

gterzian replied:
Yes that should be possible, since the set id is available at each call site I think.

rphmeier commented on core/finality-grandpa/src/communication/gossip.rs:
```diff
@@ -512,7 +560,89 @@ impl<Block: BlockT> Inner<Block> {
+		// When we're processed 2 votes of a given type,
+		// reject all further messages for this voter.
+		if should_reject {
+			return Consider::RejectRedundant
```
I'm not convinced this is the responsibility of the networking system. IMO it only serves to complicate the logic here. GRANDPA internals will already reject the redundant votes, and although it does mean that we will have to do a little extra work to get them to GRANDPA in the first place, we know that the total number of messages to process will be bounded just because we bound per-peer comms.

gterzian replied:
If we reject here they won't be stored by the networking layer. I thought that was a goal of this?

rphmeier replied:
The main goal is to make sure that the amount of data stored by the networking layer is bounded. This extra snippet means that we push the bound down a bit, but it doesn't give us an algorithmic win. (still O(n) in the number of peers).

gterzian replied:
Ok, so just to be clear, we want to remove the `Consider::RejectRedundant`, but keep the `Consider::RejectWillfulRedundant`(reporting bad peers) one?

rphmeier replied:
Yeah. But I would merge it as is if there's a good argument for reducing the bound.

gterzian replied:
I'd argue for removing it, if only because we have this equivocation api in the `Environment` and we might want to do something with it at some point, so it's probably a good idea to let those through to be handled downstream.

rphmeier commented on core/finality-grandpa/src/communication/gossip.rs:
```diff
@@ -512,7 +560,89 @@ impl<Block: BlockT> Inner<Block> {
+		let outgoing_tally_for_round = self.outgoing_votes_tally
+			.get_mut(&round)
+			.expect("If incoming votes knows about this round, so should outgoing ones; qed");
+		let outgoing_tally = outgoing_tally_for_round.entry((msg.id.clone(), who.clone())).or_insert(Default::default());
```
At this stage, the `msg.id` hasn't actually been checked to see if it's really part of the authority set at that round. That makes this a DoS vector.

gterzian replied:
Ok, I've moved the check up in trying to address this.

rphmeier commented on core/finality-grandpa/src/communication/gossip.rs:
```diff
@@ -762,55 +895,141 @@ impl<Block: BlockT> network_gossip::Validator<Block> for GossipValidator<Block>
-			let do_rebroadcast = if now >= inner.next_rebroadcast {
-				inner.next_rebroadcast = now + REBROADCAST_AFTER;
+			let do_rebroadcast = if now >= inner.read().next_rebroadcast {
+				(*inner.write()).next_rebroadcast = now + REBROADCAST_AFTER;
```
Why make this update non-atomic? now it can race and `next_rebroadcast` can move backwards in time. This is a bit overzealous on the "reduce lock hold times" front because it actually breaks an invariant...

I'd recommend acquiring a read lock and then _upgrading_ it, not releasing and re-acquiring.

gterzian replied:
Maybe there could be a tiny race, although I think it's hard to argue for it since any writing thread will first have assinged a value to `now`, then acquired the read lock, and then only acquired the write lock. So it's almost like a queue where it would be very hard for a parallel `now`to be written after the `read` that would be higher than the `now` of the current thread. 

Any way, `REBROADCAST_AFTER` is set to 5 minutes. Any potential race will be on the order or milliseconds and absolutely irrelevant. 

Finally, there is no parallelism involved in this operation, however the grandpa worker might still be waiting on the lock and happy to be able to quickly acquire and release it in between the `read` and `write` done here.


Off-course if code is written elsewhere that could hold this lock for a long time, we might be in trouble indeed. That's actually an argument for adopting this style elsewhere, and not have threads hold on to their locks to try to simulate atomic operations or transactions of some kind.  

"releasing and re-acquiring" is what parking lot is written for.

rphmeier replied:
> "releasing and re-acquiring" is what parking lot is written for.

Really tenuous claim. This is all micro-optimization anyway, and it's not worth sacrificing invariants about control flow for it.

rphmeier commented on core/finality-grandpa/src/communication/gossip.rs:
```diff
@@ -762,55 +895,141 @@ impl<Block: BlockT> network_gossip::Validator<Block> for GossipValidator<Block>
-				None => return false,
-				Some(x) => x,
+			let (maybe_round, set_id) = {
+				let inner = inner.read();
```
It doesn't make sense to re-acquire the lock (multiple times?????) on every message. Remember that this `Box<Fn>` will be called in a loop. Would you really rather write

```rust
for message in messages {
    inner.read().should_we_send(message)?;
}
```

instead of 

```rust
let inner = inner.read();
for message in messages {
    inner.should_we_send(message)?;
}
```

Again, this is jumping the gun big-time on the "minimize lock hold times" dogma. We haven't shown that this is actually a bottleneck so maintaining atomicity and consistency between message evaluations is much more important. And I'm not even convinced that your micro-optimization is even an optimization. You've set it up so it's going to do a ton more atomic ops and blow the CPU caches with fences in a tight loop.

gterzian replied:
> It doesn't make sense to re-acquire the lock (multiple times?????) on every message. 

That's exactly what `parking_lot` is designed for. 

So yes, acquiring the read lock inside the loop at each iteration is superior, or at least that is what `parking_lot` is designed for. 

The goal is not minimizing the amount of times you (re-)acquire the lock, it's minimizing the chances of a thread being parked(and a thread being parked is particularly bad if that thread is supposed to run non-blocking tasks). 

So by acquiring and releasing the locks as often as possible, you reduce the chance of a thread being parked, since it will have a chance to to a quick acquire/release of the lock in between your own operations.

rphmeier replied:
It's not what parking_lot was designed for. `parking_lot` is designed to make this case better, but it doesn't have any goal of forcing you to write code this way. 

The invariant that the state of inner doesn't change in the middle of checking if a single message should be sent to a peer is valuable and it shouldn't be thrown out for micro-optimization.

gterzian replied:
> so maintaining atomicity and consistency between message evaluations is much more important

The question is: why? What explicit logic dictates that for the entire list of messages that are being evaluated, we need `inner` to remain locked by the current thread? Why does the entire loop need to be a critical section?

On the other hand, I can imagine a grandpa voter noting a round in between, which could make several messages irrelevant, at which point we don't want to allow them anymore. Fine-grain locking allows for such "updates" in between evaluation of messages, and I think they're potentially beneficial. 

So I'd say to hold locks for as little as possible, preferably just for a single operation, unless explicit business logic requires a large critical section. If anything, minimizing the scopes of locks held will also make it harder to have unexpected deadlocks.

rphmeier replied:
> What explicit logic dictates that for the entire list of messages that are being evaluated, we need inner to remain locked by the current thread.

There isn't any. I think it's reasonable to acquire the lock and hold it for the processing of an entire message, but release afterwards. 

The main thing is that even though the ultra fine-grained acquire-and-release might be technically correct _at the moment_, you are leaving a lot of gaps for subtle races to break stuff in the future. The invariants of the code are much harder to quantify. It may seem like obvious optimization to you, but you need to be conscious that you are making a trade-off against all these things:
  - how long it takes to onboard someone onto this part of the code
  - how much time you have to spend on review
  - how much time it takes to debug

gterzian replied:
Fine, I'll remove those changes, they're not really part of the goal of this PR anyway. 

I do disagree with your arguments, since in my experience it's holding locks for longer than necessary, like all other anti-patterns, that is a subtle source of future problems. 

An example from this very file can be found at https://github.com/paritytech/substrate/pull/2533 and https://github.com/paritytech/substrate/pull/2537

rphmeier commented on core/finality-grandpa/src/communication/gossip.rs:
```diff
@@ -762,55 +895,141 @@ impl<Block: BlockT> network_gossip::Validator<Block> for GossipValidator<Block>
+							Some(peer) => peer.view.last_commit,
+						};
+            			// global message.
+            			let local_view = match inner.local_view {
```
what's going on with the indentation here?

rphmeier commented on core/finality-grandpa/src/communication/mod.rs:
```diff
@@ -64,7 +64,9 @@ mod cost {
 	pub(super) const MALFORMED_COMMIT: i32 = -1000;
 	pub(super) const FUTURE_MESSAGE: i32 = -500;
-	pub(super) const UNKNOWN_VOTER: i32 = -150;
+    pub(super) const UNKNOWN_VOTER: i32 = -150;
```
```suggestion
	pub(super) const UNKNOWN_VOTER: i32 = -150;
```

Please remember to use tabs. This was brought up multiple times in review for this PR.

rphmeier commented:
I'd love to see some tests too.

gterzian commented:
Thanks, I'll look at adding tests and addressing remaining issues.

gterzian commented:
Ok I need to look at bit more into the tests because right now they're actually always hitting the "unknown round" path...

andresilva commented on core/finality-grandpa/src/communication/tests.rs:
```diff
@@ -283,6 +283,219 @@ fn good_commit_leads_to_relay() {
+	let test = make_test_network()
+		.and_then(move |tester| {
+			// register a peer.
+			tester.gossip_validator.new_peer(&mut NoopContext, &id, network::config::Roles::FULL);
```
I think you'll need to call `note_set` on the validator otherwise messages will be dropped (since our local view is undefined).

gterzian replied:
isn't `note_set` called as part of `global_communication`(called in the test below)?

It seems the message is now dropped because the round is unknown...

gterzian replied:
Ok I think I found out the problem, see the last commit. It was because `note_round` will only add tallies for rounds that aren't the `Round(0)` added in `note_set`, which makes me wonder if we should add tallies as part of `note_set`...

rphmeier commented on core/finality-grandpa/src/communication/gossip.rs:
```diff
@@ -484,20 +484,20 @@ impl<Block: BlockT> Inner<Block> {
 		};
 
-		let set_id = local_view.set_id;
+        let set_id = local_view.set_id;
```
spaces again...

rphmeier commented on core/finality-grandpa/src/communication/mod.rs:
```diff
@@ -64,7 +64,7 @@ mod cost {
 	pub(super) const FUTURE_MESSAGE: i32 = -500;
-    pub(super) const UNKNOWN_VOTER: i32 = -150;
+	pub(super) const UNKNOWN_VOTER: i32 = -150;
     pub(super) const WILLFULREDUNDANT: i32 = -500;
```
this is still spaces

gterzian replied:
I'll address all of those in one big move at the end

gavofyork commented:
how is this looking? ready for merge soon?

gterzian commented:
Still one last question to iron out(cc @rphmeier @andresilva):

I've noticed that `note_set` does a `self.live_topics.push`. So since the tallies are only setup inside `note_round`, the round that is added as part of `note_set` is then unknown to the tallies, meaning any message related to it is going to be discarded. 

One solution could be adding some tallies setup in `note_set`, but I thought it might be simpler to just call `note_round` at the end of `note_set`(which will also take care of the `multicast_neighbor_packet`). 

Is there any reason why we should not be doing a call to `note_round` at the end of `note_set`, to note the `Round(0)` that is otherwise manually added to `live_topics` in `note_set`?

I've made the change at https://github.com/paritytech/substrate/pull/2479/commits/e80d9c69f9a80b595ea2c047f1b6aa5f8b07910e (sorry it also contains other changes, but I should it should be clear enough).

andresilva commented:
Round 0 will actually never be a live round, when we start a new authority set we add the information about round 0 (without any votes, https://github.com/paritytech/substrate/blob/master/core/finality-grandpa/src/lib.rs#L638), but the voter will start running at round 1. There should be no messages on the network related to round 0.

gterzian commented:
Ok this one is ready for review @rphmeier @andresilva 

The problem with the tests earlier seems to be that I was actually using the `Round(0)` in those. 

By the way, in the light of this `Round(0)` never being a live round, and not being part of tallied round, one can wonder if that opens a loophole in terms of Dos protection. While the fact that this round hasn't been noted means any message will be immediately discarded, the other side of that is that we won't report a peer that could keep sending us messages for that round. We won't handle those messages, but in theory many messages could still keep us busy if only at the early stage of validation, unless there is another protection at the networking layer, or we start tallying messages for that round too.

gterzian commented:
by the way, what I said above isn't really specific to `Round(0)`, it goes for any round that hasn't been currently noted, and finally, we do pass it to the `local_view` for validation, which reports a peer sending us messages that are unexpected...

rphmeier commented:
@gterzian There is an associated cost for sending past rounds' messages. So spam from Round(0), just like spam from any old round, will lead to disconnection. If we're at Round(1_000_000), messages from Round(100) are just as useless as messages from Round(0).

andresilva commented on core/finality-grandpa/src/communication/gossip.rs:
```diff
@@ -97,6 +98,17 @@ enum Consider {
+/// used both to tally incoming votes we receive and outgoing votes we sent out.
+#[derive(Debug, Default)]
+pub(crate) struct VoteTally {
+	pre_commits: usize,
```
nitpick, feel free to ignore, but everywhere else we use `prevotes` and `precommits`.

andresilva commented on core/finality-grandpa/src/communication/gossip.rs:
```diff
@@ -512,7 +546,58 @@ impl<Block: BlockT> Inner<Block> {
+				// We don't know about this round,
+				// let the local-view handle it and likely treat it as past or future.
+				return self.local_view.as_ref().map(|v| v.consider_vote(round, set_id))
+					.unwrap_or(Consider::RejectOutOfScope)
```
```suggestion
					.unwrap_or(Consider::RejectOutOfScope);
```

andresilva commented on core/finality-grandpa/src/communication/gossip.rs:
```diff
@@ -512,7 +546,58 @@ impl<Block: BlockT> Inner<Block> {
+		let outgoing_tally = outgoing_tally_for_round.entry((msg.id.clone(), who.clone())).or_insert(Default::default());
+
+		match &msg.message {
+			PrimaryPropose(_propose) => {
```
I think right now we allow multiple proposals as long as they are from the primary. Not sure if we should have this strict check here. Honest primaries should only propose once though. cc @rphmeier

rphmeier replied:
the code committed appears to be right

andresilva commented on core/finality-grandpa/src/communication/gossip.rs:
```diff
@@ -512,7 +546,58 @@ impl<Block: BlockT> Inner<Block> {
+		// We report peers who send us a third or more of message of any kind,
+		// when we've already exchanged two messages of any kind for that voter to that peer.
+		let should_report = (outgoing_tally.primary_proposals + per_peer_tally.primary_proposals) > 2 ||
+				(outgoing_tally.pre_votes + per_peer_tally.pre_votes) > 2 ||
```
```suggestion
			(outgoing_tally.pre_votes + per_peer_tally.pre_votes) > 2 ||
```

andresilva replied:
Are we sure the peer won't send us the same message we sent him?

gterzian replied:
If the peer is running this code, there is a check at `message_allowed` that is essentially the mirror or this one, so the idea is that we won't send out what we wouldn't accept, based on what we've received and sent out so far. 

So I think so, yes. Do you have something else in mind?

andresilva commented on core/finality-grandpa/src/communication/gossip.rs:
```diff
@@ -512,7 +546,58 @@ impl<Block: BlockT> Inner<Block> {
+		// when we've already exchanged two messages of any kind for that voter to that peer.
+		let should_report = (outgoing_tally.primary_proposals + per_peer_tally.primary_proposals) > 2 ||
+				(outgoing_tally.pre_votes + per_peer_tally.pre_votes) > 2 ||
+				(outgoing_tally.pre_commits + per_peer_tally.pre_commits) > 2;
```
```suggestion
			(outgoing_tally.pre_commits + per_peer_tally.pre_commits) > 2;
```

andresilva replied:
Same as above.

andresilva commented on core/finality-grandpa/src/communication/mod.rs:
```diff
@@ -65,6 +65,7 @@ mod cost {
 	pub(super) const MALFORMED_COMMIT: i32 = -1000;
 	pub(super) const FUTURE_MESSAGE: i32 = -500;
 	pub(super) const UNKNOWN_VOTER: i32 = -150;
+	pub(super) const WILLFULREDUNDANT: i32 = -500;
```
```suggestion
	pub(super) const WILLFUL_REDUNDANT: i32 = -500;
```

gterzian commented:
@andresilva Ok I have addressed your comments. Re your question at https://github.com/paritytech/substrate/pull/2479#discussion_r301329227, the idea is that this only reject what it would not allow in `message_allowed`, so we apply the same checks in what we sent out to what we receive.

gterzian commented:
Ok, I think all comments have been addressed @rphmeier @andresilva , there were a couple of CI failures however they appear unrelated, to me least:

```
error[E0308]: mismatched types
   --> /builds/parity/substrate/srml/support/procedural/src/storage/transformation.rs:955:35
    |
955 |         let type_infos = get_type_infos(storage_type);
    |                                         ^^^^^^^^^^^^
    |                                         |
    |                                         expected reference, found enum `storage::DeclStorageType`
    |                                         help: consider borrowing here: `&storage_type`
    |
    = note: expected type `&storage::DeclStorageType`
               found type `storage::DeclStorageType`
```

```
error[E0433]: failed to resolve: unresolved import
  --> /ci-cache/substrate/cargo/test-linux-stable-int/registry/src/github.com-1ecc6299db9ec823/serde_derive-1.0.94/src/fragment.rs:58:18
   |
58 |                 <Token![,]>::default().to_tokens(out);
   |                  ^^^^^^^^^
   |                  |
   |                  unresolved import
   |                  help: a similar path exists: `syn::token`
   |
   = note: this error originates in a macro outside of the current crate (in Nightly builds, run with -Z external-macro-backtrace for more info)
```

andresilva commented on core/finality-grandpa/src/communication/gossip.rs:
```diff
@@ -1090,45 +1175,127 @@ impl<Block: BlockT> network_gossip::Validator<Block> for GossipValidator<Block>
-				Some(ref v) => v,
-				None => return false, // cannot evaluate until we have a local view.
-			};
+					let (primary_proposals_to_peer, prevotes_to_peer, precommits_to_peer) = {
```
The `message_allowed` method may be called more than once with the same message, so we can't update the tallies the way we're doing here. All of the messages we see here we should have already seen through `validate` previously, except for the messages we send ourselves (which we will see here but not in `validate`).

Ideally we wouldn't need to do anything with the tallies here, but we have to update the tallies with the messages we send ourselves.

gterzian replied:
Yes, in `message_allowed` we only tally what we sent out, and in `validate` we only tally what we receive. 

> The message_allowed method may be called more than once with the same message

I think the issue here might be that it's used as part of `validate` to "broadcast topics", and then it's called as part of `propagate` in various places. 

Do you mean that in some cases we shouldn't tally what we send out, for example when broadcasting as part of `validate`. 

How can `message_allowed` "known" when an outgoing message should be tallied(when it comes from the node itself?), and when it shouldn't be tallied?

gterzian commented:
In the light of https://github.com/paritytech/substrate/pull/2479#discussion_r308392270

What can we do to get this operational? I'm under the impression that we should not tally some outgoing messages, and I'm not sure which one. The ones that are part of a (re-)broadcast we "didn't send ourselves"?

@andresilva

bkchr commented:
@gterzian what is the status of this?
